### ISSUE TITLE
Data loss when editing a chapter for several hours

### ISSUE THREAD

**Issue by rbohrer** [amyjko/bookish#340]
Account email: rose.bohrer.cs@gmail.com
I was editing chapter 1 of my book (Human Centered Programming Languages) for about two hours and, when I finished writing it, the "Saving" icon started running. While it was saving, I right-clicked to open the main menu again in a new tab. When I did this (whether the two are linked, it's unclear), my original tab flashed backed to the old draft of the chapter, with several hours of work undone. The editor then saved this old copy, resulting in data loss.

**Comment by amyjko** [comment]
Thanks for reporting Rose! Just to make sure I have the reproduction steps right:

* open chapter 1 page
* edit chapter 1 text
* open table of contents in new tab via the chapter 1 tab's table of contents link while saving
* chapter 1 text reverts to a previous version in the original tab

Does that sound correct?

I have some ideas about potential causes; I have built locking mechanisms to avoid collaborators clobbering each other's edits, but it's possible those locking mechanisms don't prevent specific author from clobbering their own edits when editing in multiple tabs. I'll try to reproduce based on the above.

**Comment by rbohrer** [comment]
I believe I specifically clicked this "Home" link, other than that, yes it's correct. Your assessment sounds right to me, but I'll comment that part of what got me was that the reverted version ended up getting autosaved, and that there wasn't an easy way to dig back through previous versions again after-the-fact. I'm not implying any particular resolution, but just providing the raw data for you do with as you wish.

(On the bright side, I have now rewritten the lost material)

![image](https://github.com/amyjko/bookish/assets/1960801/3c098c7b-c577-4368-beb6-4243de3752f6)

**Comment by rbohrer** [comment]
Alright, I have more helpful information now! I reproduced the issue while editing my next chapter. I've attached the error log from the developer tools:
[bookish.press-1685551849400.log](https://github.com/amyjko/bookish/files/11616724/bookish.press-1685551849400.log)

The most relevant error messages are the ones shown in this screenshot:
![image](https://github.com/amyjko/bookish/assets/1960801/bfd2a732-f19d-4aea-9a10-fde422e664d6)

If I had to guess what's happening:

- The editor tries to autosave very frequently so you don't lose things
- It doesn't want everything to hang all the time so it does that asynchronously and queues up writes
- All of a sudden there's a lot of writes queued up (or otherwise bunched closed together temporally) and they start getting rejected
- Client receives a 403 as a result and hangs on the "saving" step in the UI

I suppose potential routes for a fix include reducing the frequency of database calls and/or raising limits

**Comment by rbohrer** [comment]
(also, the first line of the error log indicates some smaller bug regarding indentation of bulleted lists, but I don't want to cram two bugs into one issue, and that honestly a much lower priority)

**Comment by amyjko** [comment]
This is excellent, thank you so much for sharing this. Firebase definitely has some unpredictable write behavior, but it's possible I'm not handling this particular error correctly. I'll investigate.

One question: when you reproduced this, was it while editing in multiple tabs, per the original reproduction steps, or just a single tab? (The multiple tabs problem is an issue, #295, but this may be separate).

**Comment by amyjko** [comment]
I believe I've resolved this. I couldn't reproduce, so I'm not 100% certain, but two issues were likely interacting:

* There are a range of cases that can lead to writes to failing (the write stream queue issue above, which can happen when an internet connection is interrupted, pooling writes). They're rare, but happen by design.

* When a write fails, the query snapshot wasn't guarding against pending local writes when updating chapter text from the server, and so the version in the database was overwriting the pending writes.

To resolve this, I 1) reduced the frequency of writes, which should reduce the likelihood of pooling, and 2) guarded against pending local writes when updating chapter text from the database.

There is a chance that these aren't the failure originally reported. I'll close for now, but reopen if it happens again, and I'll look for other possible causes.

**Comment by rbohrer** [comment]
It was in a single thread.  Thanks so much for the quick fix.  Will update
if it happens again

**Comment by rbohrer** [comment]
At 11:29 today I experienced the same hang-while-saving that leads to data loss, albeit with a new error log, attached
[bookish.press-1685978986272.log](https://github.com/amyjko/bookish/files/11654206/bookish.press-1685978986272.log)

**Comment by rbohrer** [comment]
After encountering the error above, I refreshed the page and got a prompt to sign in. Is it possible that my session expired while writing (which does not involve any page refreshes, because everything happens within the editor page)

**Comment by rbohrer** [comment]
I am now confident that session timeouts are a cause of data losses (not necessarily the sole cost). I had the editor open and had successfully saved all my changes so far. I left for lunch and when I came back, any further changes would leave it stuck hanging in "saving" mode. Refreshing and logging back in returned things to normal.

Potential fixes, offered with deference:

- Treat the session as active whenever the user is writing in the editor, i.e., ensure that sessions are extended while the software is in use, regardless of new page loads
- Make sessions last ~16 hours so that it is unlikely a person needs to refresh it during the day, but there is a good chance it will need refreshing at the start of the next day instead of halfway through it
- If a session is about to expire, save all changes and open a dialog in the editor that tells the user they need to log back in, instead of letting them continue typing without the ability to save it

**Comment by amyjko** [comment]
Thanks Rose, this is almost certainly a problem. Investigating a bit, it looks like Firebase Auth tokens only last an hour, and there's some cryptic process by which they get refreshed behind the scenes. That would explain why requests start coming back with 503's after a period of time. I'll go learn more.

**Comment by amyjko** [comment]
A few things I learned (just documenting them here).

The `/write` route was supposed to be reacting to changes to the auth status, but wasn't reacting to changes to the auth store. I fixed that. Any route or subroute of `/write` should now redirect to the login page with Firebase Auth expires a session. That should prevent cases where the editor appears to be operational, but actually isn't, since auth has expired. I'll release this in a moment, but it should be a rare issue.

Re: session expirations, the most useful insights I found were here from a Firebase engineer:

https://stackoverflow.com/questions/66575452/how-do-i-use-a-firebase-refresh-token-to-persist-authentication

Firebase Auth tokens expire after 1 hour, but the client side Auth SDK automatically refreshes it before it expires, so sessions should be perpetual. They're only expired when an account is deleted, disabled, or its password or email changes. So there must be some other cause of those server side request failures after a period of time.

I waited an hour to reproduce this, and as it turns out, the API that refreshes tokens (Token Service API) was restricted in production, which was causing the 503's. Because this is only used on token refresh, and not login, it only happens after an hour. This prevented the tokens from refreshed, and therefore prevented reauthentication, and therefore prevented database writes. The broken login redirect above then led to data loss, since writes weren't able to save.

https://stackoverflow.com/questions/48164248/bad-requests-in-token-service-api-when-using-firebase-auth

This is no longer restricted in production and should no longer happen. (I tested in production, and it no longer happens for me).
